safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,155 @@
+
+import (
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// LoggingService is a user-provided service for logging Fetch Metadata policy
+// violations
+type LoggingService interface {
+	Log(ir *safehttp.IncomingRequest)
+}
+
+// Plugin implements Fetch Metadata functionality.
+// See https://www.w3.org/TR/fetch-metadata/ for more details.
+type Plugin struct {
+	// Policy is the Fetch Metadata policy that decides whether a request should
+	// be allowed to pass or be rejected. This will be set to a default Resource
+	// Isolation Policy, but can be changed by the user.
+	//
+	// See https://web.dev/fetch-metadata/ for more information on the default policy
+	Policy func(*safehttp.IncomingRequest, map[string]bool) bool
+	// NavIsolation indicates whether the Navigation Isolation Policy should
+	// also be applied to the request as an additional layer of checking. This
+	// provides a way to mitigate clickjacking and reflected XSS by rejecting
+	// all cross-site navigations unless coming from origins where CORS is
+	// explicitly enabled. This can be used independently from a Resource
+	// Isolation Policy.
+	//
+	// WARNING: This is still an experimental feature and will be disabled by default.
+	NavIsolation bool
+	reportOnly   bool
+	logger       LoggingService
+	allowedCORS  map[string]bool
+}
+
+// NewPlugin creates a new Fetch Metadata plugin in enforce mode using the
+// defaultPolicy and a set of user-provided origins on which CORS is enabled.
+func NewPlugin(origins ...string) *Plugin {
+	m := map[string]bool{}
+	for _, origin := range origins {
+		m[origin] = true
+	}
+	return &Plugin{
+		allowedCORS: m,
+		Policy:      defaultPolicy,
+	}
+}
+
+func defaultPolicy(r *safehttp.IncomingRequest, allowedCORS map[string]bool) bool {
+	h := r.Header
+	switch h.Get("Sec-Fetch-Site") {
+	case "":
+		// Fetch Metadata is not supported by the browser so we allow the
+		// request to pass.
+		return true
+	case "same-origin", "same-site", "none":
+		// The request originated from a site served by your own server
+		// ("same-origin"), a subdomain of your site ("same-site", e.g.
+		// bar.foo.com made a request to foo.com) or was caused by the user
+		// explicitly interacting with the user-agent ("none"). Therefore it is
+		// allowed to pass.
+		return true
+	}
+
+	if m := r.Method(); h.Get("Sec-Fetch-Mode") == "navigate" && (m == safehttp.MethodGet || m == safehttp.MethodHead) {
+		if dest := h.Get("Sec-Fetch-Dest"); dest == "object" || dest == "embed" {
+			// The request is cross-site and originates from <object> or <embed>
+			// so it is rejected.
+			return false
+		}
+		// The request is cross-site, but a simple top-level navigation so we
+		// allow it to pass.
+		return true
+	}
+
+	origin := h.Get("Origin")
+	if _, ok := allowedCORS[origin]; ok && h.Get("Sec-Fetch-Mode") == "cors" {
+		// The request is cross-site but sent from an
+		// origin where Cross-Origin Resource Sharing is explicitly allowed.
+		return true
+	}
```

Comment:
I think this check should be the first thing we do: if an endpoint is CORS-allowed we should bail out without doing any checks.